project-koku/koku
# COST-443: COST-438: Parquet summary trigger

## Summary
- This adds the class infrastructure for processing + summarizing parquet files for each source type.
- This adds AWS summarization logic via Presto, inserting into the AWS daily summary table in Postgres
- Bills are created in Postgres when a new billing period is processed.
- Postgres daily summary partitions are recorded and created for each new billing period.

## Testing

1. Add the following env vars and or to .env and source the file 

```
ENABLE_PARQUET_PROCESSING=True
S3_BUCKET_NAME=koku-bucket
S3_ENDPOINT=http://kokuminio:9000
S3_ACCESS_KEY=kokuminioaccess
S3_SECRET=kokuminiosecret
```

2. `make docker-up-min-presto`
3. `make create-test-customer`
4. `make load-test-customer-data`
5. `docker logs koku_worker -f` to follow along
6. When done inspect http://127.0.0.1:9000/minio/koku-bucket/ for CSV and Parquet data for AWS, Azure, and OpenShift
7. Check AWS report + tag APIs for data. 


```
GET /api/cost-management/v1/reports/aws/costs/?filter[time_scope_value]=-1
HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 1,
        "compute_count": false,
        "check_tags": false,
        "filter": {
            "time_scope_value": "-1"
        },
        "group_by": {},
        "order_by": {},
        "total": {
            "infrastructure": {
                "raw": {
                    "value": 441.388253305,
                    "units": "USD"
                },
                "markup": {
                    "value": 15.538825324,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 456.927078629,
                    "units": "USD"
                }
            },
            "supplementary": {
                "raw": {
                    "value": 0.0,
                    "units": "USD"
                },
                "markup": {
                    "value": 0.0,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 0.0,
                    "units": "USD"
                }
            },
            "cost": {
                "raw": {
                    "value": 441.388253305,
                    "units": "USD"
                },
                "markup": {
                    "value": 15.538825324,
                    "units": "USD"
                },
                "usage": {
                    "value": 0.0,
                    "units": "USD"
                },
                "total": {
                    "value": 456.927078629,
                    "units": "USD"
                }
            }
        }
    },
    "links": {
        "first": "/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/reports/aws/costs/?filter%5Btime_scope_value%5D=-1&limit=100&offset=0"
    },
    "data": [
        {
            "date": "2020-10",
            "values": [
                {
                    "date": "2020-10",
                    "source_uuid": [
                        "70a9d1cf-a710-4521-8270-1f41707729e1",
                        "b952c8c5-25af-494c-93f3-8a6a7810e9e2"
                    ],
                    "infrastructure": {
                        "raw": {
                            "value": 441.388253305,
                            "units": "USD"
                        },
                        "markup": {
                            "value": 15.538825324,
                            "units": "USD"
                        },
                        "usage": {
                            "value": 0.0,
                            "units": "USD"
                        },
                        "total": {
                            "value": 456.927078629,
                            "units": "USD"
                        }
                    },
                    "supplementary": {
                        "raw": {
                            "value": 0.0,
                            "units": "USD"
                        },
                        "markup": {
                            "value": 0.0,
                            "units": "USD"
                        },
                        "usage": {
                            "value": 0.0,
                            "units": "USD"
                        },
                        "total": {
                            "value": 0.0,
                            "units": "USD"
                        }
                    },
                    "cost": {
                        "raw": {
                            "value": 441.388253305,
                            "units": "USD"
                        },
                        "markup": {
                            "value": 15.538825324,
                            "units": "USD"
                        },
                        "usage": {
                            "value": 0.0,
                            "units": "USD"
                        },
                        "total": {
                            "value": 456.927078629,
                            "units": "USD"
                        }
                    }
                }
            ]
        }
    ]
}
```

```
GET /api/cost-management/v1/tags/aws/?filter[time_scope_value]=-1
HTTP 200 OK
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: X_RH_IDENTITY, Accept

{
    "meta": {
        "count": 5,
        "filter": {
            "time_scope_value": "-1",
            "time_scope_units": "month",
            "resolution": "monthly"
        },
        "key_only": false,
        "group_by": {},
        "order_by": {},
        "access": {}
    },
    "links": {
        "first": "/api/cost-management/v1/tags/aws/?filter%5Btime_scope_value%5D=-1&limit=100&offset=0",
        "next": null,
        "previous": null,
        "last": "/api/cost-management/v1/tags/aws/?filter%5Btime_scope_value%5D=-1&limit=100&offset=0"
    },
    "data": [
        {
            "key": "app",
            "values": [
                "analytics",
                "catalog",
                "cost"
            ]
        },
        {
            "key": "environment",
            "values": [
                "dev"
            ]
        },
        {
            "key": "openshift_node",
            "values": [
                "ocp-nod"
            ]
        },
        {
            "key": "storageclass",
            "values": [
                "bravo",
                "charlie",
                "delta",
                "epsilon",
                "gamma"
            ]
        },
        {
            "key": "version",
            "values": [
                "alpha",
                "beta",
                "gamma",
                "master",
                "prod"
            ]
        }
    ]
}
```
